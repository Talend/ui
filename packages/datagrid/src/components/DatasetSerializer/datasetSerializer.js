import get from 'lodash/get';
import isArray from 'lodash/isArray';
import round from 'lodash/round';

import {
	NAMESPACE_INDEX,
	NAMESPACE_DATA,
	COLUMN_INDEX,
	QUALITY_KEY,
	QUALITY_INVALID_KEY,
	QUALITY_EMPTY_KEY,
	QUALITY_VALID_KEY,
} from '../../constants/';

/**
 * sanitizeAvro - remove the optional type
 *
 * @param  {object} 	avro field
 * @return {object}   return the shallow avro
 * @example
 * 	sanitizeAvro({
 *		 "name": "field0",
 *		 "doc": "Nom de la gare",
 *		 "type": [
 *			 "null",
 *			 {
 *				 "type": "string",
 *				 "dqType": "FR Commune",
 *				 "dqTypeKey": "FR_COMMUNE"
 *			 }
 *		 ],
 *		 "@talend-quality@": {
 *				 "0": 0,
 *				 "1": 38,
 *				 "-1": 62,
 *				 "total": 100
 *		 }
 *	}); {..., type: {type: "string", dqType: "FR Commune", dqTypeKey: "FR_COMMUNE"}}
 */
export function sanitizeAvro(avro) {
	if (!isArray(avro.type)) {
		return avro;
	}

	return {
		...avro,
		type: avro.type.find(subType => subType !== 'null'),
	};
}

/**
 * getType - manage the type from an AVRO type
 *
 * @param  {array|object} 	avro type
 * @return {string}      		return the type showed in the datagrid
 * @example
 * 	getType([{ type: 'string', dqType: '', dqTypeKey: '' }, 'null']); // string
 * 	getType({ type: 'string', dqType: '', dqTypeKey: '' }); // string*
 * 	getType({ type: 'string', dqType: 'Type', dqTypeKey: '' }); // Type*
 */
export function getType(type, mandatory = true) {
	if (isArray(type)) {
		const notNullType = type.find(subType => subType !== 'null');
		const nullType = type.find(subType => subType === 'null');

		if (notNullType && nullType) {
			return `${getType(notNullType, false)}`;
		}
	}

	return `${type.dqType || type.type}${mandatory ? '*' : ''}`;
}

export function getQuality(qualityTotal, rowsTotal) {
	return {
		percentage: rowsTotal ? round((qualityTotal / rowsTotal) * 100) : 0,
		total: qualityTotal,
	};
}

export function getFieldQuality(quality) {
	if (!quality) {
		return {};
	}

	return {
		[QUALITY_INVALID_KEY]: getQuality(quality[QUALITY_INVALID_KEY], quality.total),
		[QUALITY_EMPTY_KEY]: getQuality(quality[QUALITY_EMPTY_KEY], quality.total),
		[QUALITY_VALID_KEY]: getQuality(quality[QUALITY_VALID_KEY], quality.total),
	};
}

export function convertSample(sample) {
	if (sample.toJS) {
		return sample.toJS();
	}

	return sample;
}

export function getColumnDefs(sample) {
	if (!sample) {
		return [];
	}

	const plainObjectSample = convertSample(sample);

	return get(plainObjectSample, 'schema.fields', []).map(avroField => ({
		avro: sanitizeAvro(avroField),
		field: `${NAMESPACE_DATA}${avroField.name}`,
		headerName: avroField.doc,
		type: getType(avroField.type),
		[QUALITY_KEY]: getFieldQuality(avroField[QUALITY_KEY]),
	}));
}

export function getRowData(sample, startIndex = 0) {
	if (!sample) {
		return [];
	}

	const plainObjectSample = convertSample(sample);

	return get(plainObjectSample, 'data', []).map((row, index) =>
		Object.keys(row.value).reduce(
			(rowData, key) => ({
				...rowData,
				[`${NAMESPACE_DATA}${key}`]: {
					value: row.value[key].value,
					quality: row.value[key].quality,
					comments: [],
					avro: {},
				},
			}),
			{
				[`${NAMESPACE_INDEX}${COLUMN_INDEX}`]: index + startIndex,
				loading: !!row.loading,
			},
		),
	);
}

export function getPinnedColumnDefs(sample) {
	if (!sample) {
		return [];
	}

	return [
		{
			field: `${NAMESPACE_INDEX}${COLUMN_INDEX}`,
			width: 100,
		},
	];
}

export function getCellValue({ colDef, data }) {
	return data[colDef.field];
}
