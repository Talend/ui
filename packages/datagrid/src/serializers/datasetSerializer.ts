import { ColDef, GetRowIdFunc, ValueGetterFunc } from 'ag-grid-community';

import {
	QUALITY_EMPTY_KEY,
	QUALITY_INVALID_KEY,
	QUALITY_KEY,
	QUALITY_VALID_KEY,
} from '../constants';
import { DefaultColDef, DefaultPinnedColDef } from '../constants/column-definition.constants';
import { AvroField, Schema } from '../types';

/**
 * check if the type is null
 *
 * @param  {string|object} type  type to check
 * @return {boolean} return if the type is null
 */
function isNull(type: AvroField['type']) {
	if (typeof type === 'string') {
		return type === 'null';
	}

	return type.type === 'null';
}

/**
 * remove the optional type
 *
 * @param  {object} 	avro field
 * @return {object}   return the shallow avro
 * @example
 * 	sanitizeAvro({
 *		 "name": "field0",
 *		 "doc": "Nom de la gare",
 *		 "type": [
 *			 "null",
 *			 {
 *				 "type": "string",
 *				 "dqType": "FR Commune",
 *				 "dqTypeKey": "FR_COMMUNE"
 *			 }
 *		 ],
 *		 "@talend-quality@": {
 *				 "0": 0,
 *				 "1": 38,
 *				 "-1": 62,
 *				 "total": 100
 *		 }
 *	}); {..., type: {type: "string", dqType: "FR Commune", dqTypeKey: "FR_COMMUNE"}}
 */
export function sanitizeAvro(avroField: AvroField) {
	const rawType = Array.isArray(avroField.type)
		? avroField.type.find(it => !isNull(it))
		: avroField.type;

	return typeof rawType === 'string'
		? {
				type: rawType,
				typeLabel: rawType,
		  }
		: {
				type: rawType.type,
				typeLabel: rawType.type,
				semanticTypeId: rawType.dqTypeKey,
				semanticTypeLabel: rawType.dqType,
				logicalType: rawType.logicalType,
		  };
}

/**
 * Extract the quality from the type.
 * @param {object or array} type
 */
export function getQualityValue(type: AvroField['type']) {
	if (Array.isArray(type)) {
		return type.find(value => value[QUALITY_KEY] !== undefined)?.[QUALITY_KEY];
	}
	return type?.[QUALITY_KEY];
}

const valueGetter: ValueGetterFunc = ({ data, colDef }) => data.value?.[colDef.field!];

/**
 * Return row identifier, used when updating data
 */
export const getRowId: GetRowIdFunc = params => params.data.id;

/**
 * Add a row identifier
 */
export const parseRow = (data: any, index: number) => ({
	...data,
	// getRowId needs a string
	id: index.toString(),
});

/**
 * Return column definitions for a dataset sample
 */
export function getColumnDefs(schema: Schema): ColDef[] {
	return [
		DefaultPinnedColDef,
		...schema.fields.map(avroField => {
			const type = sanitizeAvro(avroField);
			const quality = avroField.type && getQualityValue(avroField.type);
			return {
				...DefaultColDef,
				valueGetter: valueGetter,
				field: avroField.name,
				headerName: avroField.doc || avroField.name,
				cellRendererParams: {
					...type,
					avro: type,
				},
				headerComponentParams: {
					avro: type,
					...type,
					// eslint-disable-next-line @typescript-eslint/ban-ts-comment
					// @ts-ignore
					required: !Array.isArray(avroField.type) || !avroField.type.find(isNull),
					quality: quality && {
						invalid: quality[QUALITY_INVALID_KEY],
						empty: quality[QUALITY_EMPTY_KEY],
						valid: quality[QUALITY_VALID_KEY],
					},
				},
			};
		}),
	];
}
