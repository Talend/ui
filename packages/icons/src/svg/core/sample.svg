<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16">
	<path d="M5.627 8.312s-2.401 2.51-2.924 3.053c-.523.541-1.733 1.702-.67 2.766.952.952 2.119-.236 2.772-.912.654-.677 4.713-4.899 4.713-4.899l-3.891-.008z"/>
	<path d="M5.67 13.363l-.175.183c-.667.7-1.583 1.656-2.562 1.656-.459 0-.91-.2-1.382-.616-.463-.408-.699-.832-.717-1.295-.039-.99.935-1.987 1.518-2.584l.14-.144 7.813-8.1 3.178 2.798-7.813 8.102zM16 5.207L10.087 0 8.901 1.227l.791.696C8.25 3.418 2.614 9.261 1.88 10.025l-.136.14c-.67.686-1.795 1.833-1.74 3.157.025.675.356 1.3.983 1.852.623.547 1.278.826 1.947.826 1.347 0 2.405-1.109 3.178-1.918l7.984-8.28.72.634L16 5.207z"/>
</svg>
