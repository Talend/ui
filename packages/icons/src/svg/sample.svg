<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16">
    <path d="M5.6269,8.312 C5.6269,8.312 3.2259,10.822 2.7029,11.365 C2.1799,11.906 0.9699,13.067 2.0339,14.131 C2.9849,15.083 4.1519,13.895 4.8049,13.219 C5.4589,12.542 9.5179,8.32 9.5179,8.32 L5.6269,8.312 Z"/>
    <path d="M5.6703,13.3633 L5.4953,13.5463 C4.8283,14.2463 3.9123,15.2023 2.9333,15.2023 C2.4743,15.2023 2.0223,15.0023 1.5513,14.5863 C1.0883,14.1783 0.8523,13.7543 0.8343,13.2913 C0.7953,12.3003 1.7693,11.3043 2.3523,10.7073 L2.4923,10.5633 C3.2283,9.8013 8.8643,3.9563 10.3053,2.4633 L13.4833,5.2613 L5.6703,13.3633 Z M16.0003,5.2073 L10.0873,0.0003 L8.9013,1.2273 L9.6923,1.9233 C8.2503,3.4183 2.6143,9.2613 1.8793,10.0253 L1.7433,10.1653 C1.0723,10.8513 -0.0517,11.9983 0.0023,13.3223 C0.0283,13.9973 0.3593,14.6213 0.9863,15.1743 C1.6093,15.7213 2.2643,16.0003 2.9333,16.0003 C4.2803,16.0003 5.3383,14.8913 6.1113,14.0823 L14.0953,5.8013 L14.8153,6.4363 L16.0003,5.2073 Z"/>
</svg>
