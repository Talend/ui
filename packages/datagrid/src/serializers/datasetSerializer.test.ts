import sample from '../../mocks/sample.json';
import { getColumnDefs } from './datasetSerializer';

describe('getColumnDefs', () => {
	it('should returns the columns definitions', () => {
		const columnDefs = getColumnDefs(sample.schema);

		expect(columnDefs).toMatchSnapshot();
	});
});
