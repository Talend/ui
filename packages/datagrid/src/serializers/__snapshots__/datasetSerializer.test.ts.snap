// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getColumnDefs should returns the columns definitions 1`] = `
Array [
  Object {
    "headerComponent": [Function],
    "lockPosition": true,
    "pinned": "left",
    "valueGetter": [Function],
    "width": 100,
  },
  Object {
    "cellClassRules": Object {
      "column-focus": [Function],
      "highlighted": [Function],
    },
    "cellEditor": Object {
      "$$typeof": Symbol(react.forward_ref),
      "render": [Function],
    },
    "cellRenderer": [Function],
    "cellRendererParams": Object {
      "avro": Object {
        "doc": "Nom de la gare",
        "logicalType": undefined,
        "name": "field0",
        "semanticTypeId": "FR_COMMUNE",
        "semanticTypeLabel": "FR Commune",
        "talend.component.label": "Le nom de la gare",
        "type": "string",
        "typeLabel": "string",
      },
      "doc": "Nom de la gare",
      "logicalType": undefined,
      "name": "field0",
      "semanticTypeId": "FR_COMMUNE",
      "semanticTypeLabel": "FR Commune",
      "talend.component.label": "Le nom de la gare",
      "type": "string",
      "typeLabel": "string",
    },
    "field": "field0",
    "headerClass": [Function],
    "headerComponent": [Function],
    "headerComponentParams": Object {
      "avro": Object {
        "doc": "Nom de la gare",
        "logicalType": undefined,
        "name": "field0",
        "semanticTypeId": "FR_COMMUNE",
        "semanticTypeLabel": "FR Commune",
        "talend.component.label": "Le nom de la gare",
        "type": "string",
        "typeLabel": "string",
      },
      "doc": "Nom de la gare",
      "logicalType": undefined,
      "name": "field0",
      "quality": Object {
        "empty": 0,
        "invalid": 62,
        "valid": 38,
      },
      "required": false,
      "semanticTypeId": "FR_COMMUNE",
      "semanticTypeLabel": "FR Commune",
      "talend.component.label": "Le nom de la gare",
      "type": "string",
      "typeLabel": "string",
    },
    "headerName": "Nom de la gare",
    "minWidth": 30,
    "resizable": true,
    "valueGetter": [Function],
    "width": 150,
  },
  Object {
    "cellClassRules": Object {
      "column-focus": [Function],
      "highlighted": [Function],
    },
    "cellEditor": Object {
      "$$typeof": Symbol(react.forward_ref),
      "render": [Function],
    },
    "cellRenderer": [Function],
    "cellRendererParams": Object {
      "avro": Object {
        "doc": undefined,
        "name": "field1",
        "talend.component.label": undefined,
        "type": "int",
        "typeLabel": "int",
      },
      "doc": undefined,
      "name": "field1",
      "talend.component.label": undefined,
      "type": "int",
      "typeLabel": "int",
    },
    "field": "field1",
    "headerClass": [Function],
    "headerComponent": [Function],
    "headerComponentParams": Object {
      "avro": Object {
        "doc": undefined,
        "name": "field1",
        "talend.component.label": undefined,
        "type": "int",
        "typeLabel": "int",
      },
      "doc": undefined,
      "name": "field1",
      "quality": undefined,
      "required": false,
      "talend.component.label": undefined,
      "type": "int",
      "typeLabel": "int",
    },
    "headerName": "field1",
    "minWidth": 30,
    "resizable": true,
    "valueGetter": [Function],
    "width": 150,
  },
  Object {
    "cellClassRules": Object {
      "column-focus": [Function],
      "highlighted": [Function],
    },
    "cellEditor": Object {
      "$$typeof": Symbol(react.forward_ref),
      "render": [Function],
    },
    "cellRenderer": [Function],
    "cellRendererParams": Object {
      "avro": Object {
        "doc": "Code postal",
        "logicalType": undefined,
        "name": "field2",
        "semanticTypeId": "FR_POSTAL_CODE",
        "semanticTypeLabel": "FR Postal Code",
        "talend.component.label": "Le code postal",
        "type": "string",
        "typeLabel": "string",
      },
      "doc": "Code postal",
      "logicalType": undefined,
      "name": "field2",
      "semanticTypeId": "FR_POSTAL_CODE",
      "semanticTypeLabel": "FR Postal Code",
      "talend.component.label": "Le code postal",
      "type": "string",
      "typeLabel": "string",
    },
    "field": "field2",
    "headerClass": [Function],
    "headerComponent": [Function],
    "headerComponentParams": Object {
      "avro": Object {
        "doc": "Code postal",
        "logicalType": undefined,
        "name": "field2",
        "semanticTypeId": "FR_POSTAL_CODE",
        "semanticTypeLabel": "FR Postal Code",
        "talend.component.label": "Le code postal",
        "type": "string",
        "typeLabel": "string",
      },
      "doc": "Code postal",
      "logicalType": undefined,
      "name": "field2",
      "quality": Object {
        "empty": 2,
        "invalid": 10,
        "valid": 88,
      },
      "required": true,
      "semanticTypeId": "FR_POSTAL_CODE",
      "semanticTypeLabel": "FR Postal Code",
      "talend.component.label": "Le code postal",
      "type": "string",
      "typeLabel": "string",
    },
    "headerName": "Code postal",
    "minWidth": 30,
    "resizable": true,
    "valueGetter": [Function],
    "width": 150,
  },
  Object {
    "cellClassRules": Object {
      "column-focus": [Function],
      "highlighted": [Function],
    },
    "cellEditor": Object {
      "$$typeof": Symbol(react.forward_ref),
      "render": [Function],
    },
    "cellRenderer": [Function],
    "cellRendererParams": Object {
      "avro": Object {
        "doc": "Segmentation",
        "logicalType": undefined,
        "name": "field3",
        "semanticTypeId": "",
        "semanticTypeLabel": "",
        "talend.component.label": undefined,
        "type": "boolean",
        "typeLabel": "boolean",
      },
      "doc": "Segmentation",
      "logicalType": undefined,
      "name": "field3",
      "semanticTypeId": "",
      "semanticTypeLabel": "",
      "talend.component.label": undefined,
      "type": "boolean",
      "typeLabel": "boolean",
    },
    "field": "field3",
    "headerClass": [Function],
    "headerComponent": [Function],
    "headerComponentParams": Object {
      "avro": Object {
        "doc": "Segmentation",
        "logicalType": undefined,
        "name": "field3",
        "semanticTypeId": "",
        "semanticTypeLabel": "",
        "talend.component.label": undefined,
        "type": "boolean",
        "typeLabel": "boolean",
      },
      "doc": "Segmentation",
      "logicalType": undefined,
      "name": "field3",
      "quality": Object {
        "empty": 0,
        "invalid": 4,
        "valid": 96,
      },
      "required": true,
      "semanticTypeId": "",
      "semanticTypeLabel": "",
      "talend.component.label": undefined,
      "type": "boolean",
      "typeLabel": "boolean",
    },
    "headerName": "Segmentation",
    "minWidth": 30,
    "resizable": true,
    "valueGetter": [Function],
    "width": 150,
  },
  Object {
    "cellClassRules": Object {
      "column-focus": [Function],
      "highlighted": [Function],
    },
    "cellEditor": Object {
      "$$typeof": Symbol(react.forward_ref),
      "render": [Function],
    },
    "cellRenderer": [Function],
    "cellRendererParams": Object {
      "avro": Object {
        "doc": "Planned date",
        "logicalType": undefined,
        "name": "field4",
        "semanticTypeId": "",
        "semanticTypeLabel": "",
        "talend.component.label": undefined,
        "type": "date",
        "typeLabel": "date",
      },
      "doc": "Planned date",
      "logicalType": undefined,
      "name": "field4",
      "semanticTypeId": "",
      "semanticTypeLabel": "",
      "talend.component.label": undefined,
      "type": "date",
      "typeLabel": "date",
    },
    "field": "field4",
    "headerClass": [Function],
    "headerComponent": [Function],
    "headerComponentParams": Object {
      "avro": Object {
        "doc": "Planned date",
        "logicalType": undefined,
        "name": "field4",
        "semanticTypeId": "",
        "semanticTypeLabel": "",
        "talend.component.label": undefined,
        "type": "date",
        "typeLabel": "date",
      },
      "doc": "Planned date",
      "logicalType": undefined,
      "name": "field4",
      "quality": Object {
        "empty": 4,
        "invalid": 0,
        "valid": 96,
      },
      "required": true,
      "semanticTypeId": "",
      "semanticTypeLabel": "",
      "talend.component.label": undefined,
      "type": "date",
      "typeLabel": "date",
    },
    "headerName": "Planned date",
    "minWidth": 30,
    "resizable": true,
    "valueGetter": [Function],
    "width": 150,
  },
  Object {
    "cellClassRules": Object {
      "column-focus": [Function],
      "highlighted": [Function],
    },
    "cellEditor": Object {
      "$$typeof": Symbol(react.forward_ref),
      "render": [Function],
    },
    "cellRenderer": [Function],
    "cellRendererParams": Object {
      "avro": Object {
        "doc": "volume total d'usagers 2016 (voyageurs+ non voyageurs)",
        "logicalType": undefined,
        "name": "field5",
        "semanticTypeId": "",
        "semanticTypeLabel": "",
        "talend.component.label": undefined,
        "type": "unknown",
        "typeLabel": "unknown",
      },
      "doc": "volume total d'usagers 2016 (voyageurs+ non voyageurs)",
      "logicalType": undefined,
      "name": "field5",
      "semanticTypeId": "",
      "semanticTypeLabel": "",
      "talend.component.label": undefined,
      "type": "unknown",
      "typeLabel": "unknown",
    },
    "field": "field5",
    "headerClass": [Function],
    "headerComponent": [Function],
    "headerComponentParams": Object {
      "avro": Object {
        "doc": "volume total d'usagers 2016 (voyageurs+ non voyageurs)",
        "logicalType": undefined,
        "name": "field5",
        "semanticTypeId": "",
        "semanticTypeLabel": "",
        "talend.component.label": undefined,
        "type": "unknown",
        "typeLabel": "unknown",
      },
      "doc": "volume total d'usagers 2016 (voyageurs+ non voyageurs)",
      "logicalType": undefined,
      "name": "field5",
      "quality": Object {
        "empty": 4,
        "invalid": 0,
        "valid": 96,
      },
      "required": true,
      "semanticTypeId": "",
      "semanticTypeLabel": "",
      "talend.component.label": undefined,
      "type": "unknown",
      "typeLabel": "unknown",
    },
    "headerName": "volume total d'usagers 2016 (voyageurs+ non voyageurs)",
    "minWidth": 30,
    "resizable": true,
    "valueGetter": [Function],
    "width": 150,
  },
  Object {
    "cellClassRules": Object {
      "column-focus": [Function],
      "highlighted": [Function],
    },
    "cellEditor": Object {
      "$$typeof": Symbol(react.forward_ref),
      "render": [Function],
    },
    "cellRenderer": [Function],
    "cellRendererParams": Object {
      "avro": Object {
        "doc": "voyageurs 2015",
        "name": "field6",
        "talend.component.label": undefined,
        "type": "bytes",
        "typeLabel": "bytes",
      },
      "doc": "voyageurs 2015",
      "name": "field6",
      "talend.component.label": undefined,
      "type": "bytes",
      "typeLabel": "bytes",
    },
    "field": "field6",
    "headerClass": [Function],
    "headerComponent": [Function],
    "headerComponentParams": Object {
      "avro": Object {
        "doc": "voyageurs 2015",
        "name": "field6",
        "talend.component.label": undefined,
        "type": "bytes",
        "typeLabel": "bytes",
      },
      "doc": "voyageurs 2015",
      "name": "field6",
      "quality": undefined,
      "required": false,
      "talend.component.label": undefined,
      "type": "bytes",
      "typeLabel": "bytes",
    },
    "headerName": "voyageurs 2015",
    "minWidth": 30,
    "resizable": true,
    "valueGetter": [Function],
    "width": 150,
  },
  Object {
    "cellClassRules": Object {
      "column-focus": [Function],
      "highlighted": [Function],
    },
    "cellEditor": Object {
      "$$typeof": Symbol(react.forward_ref),
      "render": [Function],
    },
    "cellRenderer": [Function],
    "cellRendererParams": Object {
      "avro": Object {
        "doc": "volume total d'usagers 2015 (voyageurs+ non voyageurs)",
        "logicalType": undefined,
        "name": "field7",
        "semanticTypeId": "",
        "semanticTypeLabel": "",
        "talend.component.label": undefined,
        "type": "string",
        "typeLabel": "string",
      },
      "doc": "volume total d'usagers 2015 (voyageurs+ non voyageurs)",
      "logicalType": undefined,
      "name": "field7",
      "semanticTypeId": "",
      "semanticTypeLabel": "",
      "talend.component.label": undefined,
      "type": "string",
      "typeLabel": "string",
    },
    "field": "field7",
    "headerClass": [Function],
    "headerComponent": [Function],
    "headerComponentParams": Object {
      "avro": Object {
        "doc": "volume total d'usagers 2015 (voyageurs+ non voyageurs)",
        "logicalType": undefined,
        "name": "field7",
        "semanticTypeId": "",
        "semanticTypeLabel": "",
        "talend.component.label": undefined,
        "type": "string",
        "typeLabel": "string",
      },
      "doc": "volume total d'usagers 2015 (voyageurs+ non voyageurs)",
      "logicalType": undefined,
      "name": "field7",
      "quality": Object {
        "empty": 4,
        "invalid": 0,
        "valid": 96,
      },
      "required": true,
      "semanticTypeId": "",
      "semanticTypeLabel": "",
      "talend.component.label": undefined,
      "type": "string",
      "typeLabel": "string",
    },
    "headerName": "volume total d'usagers 2015 (voyageurs+ non voyageurs)",
    "minWidth": 30,
    "resizable": true,
    "valueGetter": [Function],
    "width": 150,
  },
  Object {
    "cellClassRules": Object {
      "column-focus": [Function],
      "highlighted": [Function],
    },
    "cellEditor": Object {
      "$$typeof": Symbol(react.forward_ref),
      "render": [Function],
    },
    "cellRenderer": [Function],
    "cellRendererParams": Object {
      "avro": Object {
        "doc": "voyageurs 2014",
        "logicalType": undefined,
        "name": "field8",
        "semanticTypeId": "",
        "semanticTypeLabel": "",
        "talend.component.label": undefined,
        "type": "string",
        "typeLabel": "string",
      },
      "doc": "voyageurs 2014",
      "logicalType": undefined,
      "name": "field8",
      "semanticTypeId": "",
      "semanticTypeLabel": "",
      "talend.component.label": undefined,
      "type": "string",
      "typeLabel": "string",
    },
    "field": "field8",
    "headerClass": [Function],
    "headerComponent": [Function],
    "headerComponentParams": Object {
      "avro": Object {
        "doc": "voyageurs 2014",
        "logicalType": undefined,
        "name": "field8",
        "semanticTypeId": "",
        "semanticTypeLabel": "",
        "talend.component.label": undefined,
        "type": "string",
        "typeLabel": "string",
      },
      "doc": "voyageurs 2014",
      "logicalType": undefined,
      "name": "field8",
      "quality": Object {
        "empty": 4,
        "invalid": 0,
        "valid": 96,
      },
      "required": true,
      "semanticTypeId": "",
      "semanticTypeLabel": "",
      "talend.component.label": undefined,
      "type": "string",
      "typeLabel": "string",
    },
    "headerName": "voyageurs 2014",
    "minWidth": 30,
    "resizable": true,
    "valueGetter": [Function],
    "width": 150,
  },
  Object {
    "cellClassRules": Object {
      "column-focus": [Function],
      "highlighted": [Function],
    },
    "cellEditor": Object {
      "$$typeof": Symbol(react.forward_ref),
      "render": [Function],
    },
    "cellRenderer": [Function],
    "cellRendererParams": Object {
      "avro": Object {
        "doc": "volume total d'usagers 2014 (voyageurs+ non voyageurs)",
        "logicalType": undefined,
        "name": "field9",
        "semanticTypeId": "",
        "semanticTypeLabel": "",
        "talend.component.label": undefined,
        "type": "string",
        "typeLabel": "string",
      },
      "doc": "volume total d'usagers 2014 (voyageurs+ non voyageurs)",
      "logicalType": undefined,
      "name": "field9",
      "semanticTypeId": "",
      "semanticTypeLabel": "",
      "talend.component.label": undefined,
      "type": "string",
      "typeLabel": "string",
    },
    "field": "field9",
    "headerClass": [Function],
    "headerComponent": [Function],
    "headerComponentParams": Object {
      "avro": Object {
        "doc": "volume total d'usagers 2014 (voyageurs+ non voyageurs)",
        "logicalType": undefined,
        "name": "field9",
        "semanticTypeId": "",
        "semanticTypeLabel": "",
        "talend.component.label": undefined,
        "type": "string",
        "typeLabel": "string",
      },
      "doc": "volume total d'usagers 2014 (voyageurs+ non voyageurs)",
      "logicalType": undefined,
      "name": "field9",
      "quality": Object {
        "empty": 4,
        "invalid": 0,
        "valid": 96,
      },
      "required": true,
      "semanticTypeId": "",
      "semanticTypeLabel": "",
      "talend.component.label": undefined,
      "type": "string",
      "typeLabel": "string",
    },
    "headerName": "volume total d'usagers 2014 (voyageurs+ non voyageurs)",
    "minWidth": 30,
    "resizable": true,
    "valueGetter": [Function],
    "width": 150,
  },
]
`;
